src/katgpucbf/xbgpu/bsend.py

```diff
@@ -0,0 +1,390 @@
+
+        # Multicast stream parameters
+        self.heap_payload_size_bytes = n_channels_per_substream * spectra_per_heap * COMPLEX * SEND_DTYPE.itemsize
+        # Transport-agnostic stream information
+        # Used in XSend to calculate `send_rate_bytes_per_second`, do we need it here?
+        # TODO: Scope to move this calculation into a helper in utils
+        # packets_per_heap = math.ceil(self.heap_payload_size_bytes / packet_payload)
+        # packet_header_overhead_bytes = packets_per_heap * BSend.header_size
+
+        stream_config = spead2.send.StreamConfig(
+            max_packet_size=packet_payload + BSend.header_size,
+            max_heaps=n_tx_items * heaps_per_fengine_per_chunk + 1,  # TODO: Update this to be proper
+            rate_method=spead2.send.RateMethod.AUTO,
+            rate=0.0,  # TODO: Update to use `send_rate_bytes_per_second`, this sends as fast as possible
+        )
+        self.stream = stream_factory(stream_config, buffers)
+
+        item_group = spead2.send.ItemGroup(flavour=FLAVOUR)
+        item_group.add_item(
+            FREQUENCY_ID,
+            "frequency",  # Misleading name, but it's what the ICD specifies
+            "Value of the first channel in collections stored here.",
+            shape=[],
+            format=IMMEDIATE_FORMAT,
+        )
+        item_group.add_item(
+            TIMESTAMP_ID,
+            "timestamp",
+            "Timestamp provided by the MeerKAT digitisers and scaled to the digitiser sampling rate.",
+            shape=[],
+            format=IMMEDIATE_FORMAT,
+        )
+        item_group.add_item(
+            BF_RAW_ID,
+            "bf_raw",
+            "Channelised complex data.",  # TODO: Straight from the ICD, is this sufficient?
+            shape=buffers[0].shape[2:],
+            dtype=buffers[0].dtype,
+        )
+
+        self.descriptor_heap = item_group.get_heap(descriptors="all", data="none")
+
+    def enable_substream(self, stream_id: int = 0, enable: bool = True) -> None:
+        """Enable/Disable a substream's data transmission.
+
+        :class:`.BSend` operates as a large single stream with multiple
+        substreams. Each substream is its own data product and is required
+        to be enabled/disabled independently.
+
+        .. todo::
+            Update to actually carry out substream enable/disable once
+            multiple substreams are supported.
+        Parameters
+        ----------
+        stream_id
+            ID of the substream, corresponds to the <beam-id><pol>
+            convention, e.g. stream_id 3 has a stream_name ending in <1x>.
+        enable
+            Boolean indicating whether the `stream_id` should be enabled or
+            disabled.
+        """
+        self.tx_enabled = enable
+
+    async def get_free_chunk(self) -> Chunk:
+        """Return a Chunk once it has completed its send futures."""
+        chunk = await self._chunks_queue.get()
+        await chunk.future
```

Comment:
I'm slightly nervous about what happens if this await is cancelled - the chunk will probably get dropped on the floor, and we won't wait for it to complete transmission. On the other hand, that should only happen during shutdown, at which point we no longer care.

Reply:
This comment might get lost in the various updates, and I don't think we've explicitly discussed this. I think I'll wrap it in a `try...except...finally` like you've done in `chunk_send_and_cleanup` - also like in my previous update to `send_chunk`.